### ISSUE TITLE
:sparkles: [Modders and game devs] What's the 3 features you need the most?

### ISSUE THREAD

**Issue by Zughy** [minetest/minetest#15092]
After a discussion on IRC, I thought of running this experiment, similar to #14004 

If you're a developer and/or a server admin, list here the 3 features you need the most to turn ideas into reality. To do so, justify every point with a **short** description. I'll then go through every issue and decide whether to label it as "high priority" or not (to justify the criteria: I've been modding and owning a server for 4 years now, and I work as a Minetest game dev consultant. Minor: I'm the one with the most useful reviews on CDB).

Rules:
1. link features by following the format below
2. justify your choices, e.g. why is so important to you? Would it be a game changer in the gameplay experience etc.
3. keep it short
4. **at least one feature must not be yours**: I think it might be useful for everyone if we look around us
5. Let's avoid comments not following the format. If you want to correct what's been said by another user, it's fine. If more messages about the thing starts, it's not fine and they'll be hidden

---

Format

```
**What I do**:  
<!-- your CDB profile, MT server and/or an external link. Short descriptions such as "I'm an educator who.." are fine as well -->

**Features**:
1. link + reason 
2. "
3. "
```

**Comment by sfence** [comment]
**Mods and etc.**:  
https://content.minetest.net/users/SFENCE/

**Features**:
1. #14011 -> Because developers of some games (Exile etc.) are fighting with this and have to create crazy go-around code. Changing chance and time is not always a solution, because it also affects probability distribution, especially for multiple-step events.
2. #14231 -> Advtrains view bug: https://forum.minetest.net/viewtopic.php?t=28254&hilit=advtrains+camera+offset, plane mods from Apercy, everything mod which attaches the player to object with free move and rotation in space.

**Comment by lhofhansl** [comment]
@sfence Can you outline what exactly is the issue with ABM that requires the work-around? (maybe in #14011). I feel like there's a lot of confusion around this.

**Comment by sfence** [comment]
> @sfence Can you outline what exactly is the issue with ABMs that requires the work-arounds? (maybe in #14011). I feel like there's a lot of confusion around this.

The typical problem is that you want something to happen and be a little predictable. You can add a timer everywhere, but it can cause lags. You can use ABMs, but they can be skipped for many nodes. If the server maintainer starts to play with the server step (100/50ms) big effect on ABMs can be expected. The maintainer now has nothing to do with it without changing all ABM parameters game environment reactions, and mechanics. It can affect flora growing, lava cooling, terrain terraformation, fire behaviors, etc..
I think that making ABM spreadable over more global steps can help with this problem. As a bonus, it allows better use of the server performance in time and probably also a better player experience. The environment will react more predictably also on busy servers. 
I have noticed that the servers are doing nothing most of the time and then suddenly there is a rush of ABM calls and ABM skipping and then nothing again. And there is no performance reason for skipping them. If the remaining ABM is done in other steps it will result only in a better and more predictable environment. On the PvP server, smaller server steps can be the difference between life and death and 200ms is a long step for PvP.

**Comment by kromka-chleba** [comment]
**What I do**:  
I made multiple contributions to [Exile](https://content.minetest.net/packages/Mantar/exile/) such as 3D models, seasonal changes, performance improvements, etc.

I'm also making my own horror game called [Perfect City](https://codeberg.org/perfect_city_game_studio/perfect_city)  (very early dev stage).

**Features**:
1. https://github.com/minetest/minetest/issues/2670 it would greatly improve immersion and mood of my game, for example I could do orange lights for sodium street lights or red lights for some nightmarish locations.
1. https://github.com/minetest/minetest/issues/10647 as above, higher view range would improve immersion
1. https://github.com/minetest/minetest/issues/15014 I could be mistaken but I believe this is one of the things responsible for lags that are visible for example when server sends mapblocks to the client.

**Comment by nauta-turbidus** [comment]
@ryvnf Actually, natural light *does* have a different color than artificial. See vertex shaders for details.

**Comment by nauta-turbidus** [comment]
It's worse for shaders. They have a "day/night ratio" and infer the rest from it. I'm working on it and banging my head against a wall tbh.

**Comment by kromka-chleba** [comment]
>Can you outline what exactly is the issue with ABMs that requires the work-arounds? 

@lhofhansl, I'm the author of the crazy workaround made for Exile.

Some problems I encountered with ABMs:

1. Weak performance, if you add too much the game becomes laggy or ABMs don't do the job they should
2. Limited range to active block radius (which could be seen as a feature or a limitation depending on task)
3. Pitiful control over execution: can only register once, can't unregister or change, the only control over their execution is chance and interval.

The above makes ABMs a poor choice for not-so-abstract tasks such as snow/ice placing, soaking soil with water, changing seasons. I tried using ABMs and LBMs for these tasks in Exile, they delivered poor outcomes, e.g. changing seasonal soil, placing snow, etc. only in the active block radius around players with big intervals to prevent massive lags.

To solve these problems I was forced to write a mod called [Mapchunk Shepherd](https://codeberg.org/perfect_city_game_studio/mapchunk_shepherd) (that is terribly written and poorly documented btw, will fix this soon-ish) which is basically a system for assigning labels to mapchunks and then processing these mapchunks with Voxel Manip. The advantage of my system is that worker functions run only when they're needed (unlike ABMs) so ice freezers work only when temperature falls below zero, snow is only placed when it snows, seasonal soil and plants are only replaced in mapchunks with off-season flora. The advantage of my system is also that it works on all loaded mapchunks (mapblocks technically) so these tasks are performed up to the horizon with reasonable performance.

Overall I recommend using ABMs only for truly active nodes that require some high-level functions. Tasks such as replacing nodes or moving them in a simple manner could be performed by systems like mine.

**Comment by jeremyshannon** [comment]
**What I do**:
Lead developer of Exile. Sometimes I contribute a thing to other projects as well, and I wrote VH1, a little mod to add hp bars over mobs' and players' heads

**Features**:
1. #13046 An obvious big one, Exile needs this for better handling of controls (AUX1 in particular) The CSM I wrote is a poor substitute
2. *???*
3. #3878 Not important, but would be nice to have. I wanna sit in my cave and play a flute carved from wood or bone, and tracker music support could let me compose my own songs on the fly instead of just playing a canned song

**Comment by corarona** [comment]
mineclonia, several mods (some on cdb, most on https://codeberg.org/cora) , [my own wip game](https://codeberg.org/cora/musi_pona), oysterity anarchy server

1. https://github.com/minetest/minetest/pull/14260 - this would make lua mapgens a lot more attractive
2. https://github.com/minetest/minetest/pull/14333 
3. Also agree with what's been said about ABMs however maybe it would be easier to simply add more conditions to where ABMs apply e.g. biomes or make more complex nodename/neighbor conditions possilbe ( https://github.com/minetest/minetest/issues/13173 )

**Comment by rudzik8** [comment]
**What I do**:  
I'm a [VoxeLibre](https://content.minetest.net/packages/Wuzzy/mineclone2/) modder and developer (technically has been one since May 2022), as well as the current maintainer of [Voxelgarden](https://content.minetest.net/packages/Casimir/voxelgarden/), a game that is more than 11 years old now.

<details>
  <summary>My profiles</summary>
  
  [ContentDB](https://content.minetest.net/users/rudzik8/) | [Codeberg](https://codeberg.org/rudzik8)
  
</details>

**Features**:
1. #5332 is the only option I see to actually implement custom gamemodes that are not simply creative and non-creative. Here in Voxelgarden, the previous maintainer (@CasimirKaPazi) developed a custom mapgen/gamemode for it called [Unter Null](https://codeberg.org/voxelgarden/unternull), but due to the current limitations has to be distributed as a mod.
2. #14343 would help create pretty specific atmospheres in games with lighting environments. Both in VoxeLibre and Voxelgarden there are hellish Nether realms, although pretty different, and both would benefit from this. a) we could make them not pitch black and b) we could add a bit of a color inbalance towards red/orange to show just how hot it is.
3. #2670 is also related to light, but in a different way. It would be nice if not all light sources would shine the same way. Warm light from torches, yellow-ish from mese lamps (VG), etc. It would help create an environment that feels more alive and real.

**Comment by Xeno333** [comment]
I develop for a fork of mineclonia for my server Imperium ([xeno333.github.io/Imperium-site](url)), I also do other mods, and game dev.

1. #11446 Where to start on this one... modders can realisticly only use zoom or aux1 for anything and in practice its better to use a key combonation. It causes a lot of conflicts with other mods and/or player confusion. A dedicated mod key(s) is a MUST. 
2. #14344  Really really need this for armour bugs.
3. #14343 Lighting for caves and demensions.

**Comment by kno10** [comment]
**What I do**:  
I play with my kids, and try to fix bugs that annoy me (e.g., the collision with bouncy blocks bug #15029 that causes villagers to bounce through the roofs when there is lag).

**Features**:
1. Better pathfinding. The pathfinder of the minetest engine is very poor:
    - it only supports mobs of height 1
    - it does not work with the fences of Voxelibre, Mineclonia, etc. that are slightly taller than one block, and cannot be jumped over with jump height 1
    This causes villagers to be stuck very often, in particular at fences and various decoration because the pathfinder tells them to jump over them.
2. `get_node_raw`/`set_node_raw` API, coordinate-based APIs instead of vector-based APIs.
    - There is the minetest policy not to "clutter" an API, but I prefer to have different levels of working. Right now, we have the core `get_node` API, the voxelmanip `get_node_at` API (which is less optimized, lacking the equivalent optimizations of #14384), and the voxelmanip `get_data` API (that copies the entire area into a lua array of boxed numbers, so likely some 16 bytes per entry?). I would like to have `core.get_data_raw` a public API as well as `voxelmanip.get_node_raw`, if you need to access only a subset of the nodes in a particular area. This API allows working with node ids, without making a lua copy of the entire area as in voxelmanip.
   - There are many cases where a coordinate-based API is more elegant than a vector-based API (and more efficient).
     For example, `get_node_raw(x, y + 1, z)` vs. `get_node(vector.new(x, y + 1, z))` or `get_node(vector.offset(pos, 0, 1, 0))`). Once you need to offset positions, the vector API can become less elegant than raw coordinates (and is supposedly less efficient because of the tables).
3. better support for spawning larger structures / decorations
  Right now, mods typically hook into the decoration API, place a 1x1x1 temporary block, then try to emerge a sufficiently large area (which can cause trouble if there are many structures and these areas overlap a lot, triggering recursive emerges) to place the structures. But for very large structures and when the player is moving fast, this may affect areas close to the player.

**Comment by MisterE123** [comment]
What I do: Make minigames (for fun and [for money](https://gitlab.com/district-53-commission/fantasia-royale) and sometimes host them on servers.

1. https://github.com/minetest/minetest/issues/11640 or similar. Some way to unload media from the client's cache. My use case, one of many, is a custom HUD minimap that is generated by the server. Currently (using texmods), the client would consume something like a GB in 40 mins of gameplay or so. Other usecases include TVs, Digiscreens, and much more.

2. https://github.com/minetest/minetest/issues/10153 Voxel area entities with good collision physics. Modders (apercy, sumi, greenxenith) have already tried making large vehicles. They are simply limited by the engine. VAE + physics would be very useful for some minigame ideas as well.

3. SSCSM. I think we can all agree that this is limiting modders *alot*. FPS (Blockleague minigame, 4dguns), My Blockbomber minigame, HUDs that stay up-to-date despite lag, etc.

**Comment by ancientmarinerdev** [comment]
Former maintainer of VoxeLibre/MineClone2. Created a skyblock mod for it and have been working on a prototype tower defence game.

For me, priorities are as follows:

1) ABMs have so much potential, but most useful stuff goes on the global step as they're unreliable. They need to become less unreliable, and spreading over multiple globalsteps could help with this: https://github.com/minetest/minetest/issues/14011
2) Ambient Light (#14343) - Overall gameplay is poor in dark caves and darker dimensions like the nether. Currently there is no good way for mod developers to help improve this. Ambient light would help massively with this.
3) Still deciding. Will update soon.

**Comment by Zughy** [comment]
(@kno10 please link issues, it makes my life easier. If there is none, please create one)

**Comment by Zughy** [comment]
My two cents. I've tried to think of things games might need in general, and not features specifically for my content:
1. #10416 : in order to customise the experience, we can't keep going on with the default font. I can't name all the times that the default font has ruined a well designed UI/HUD
2. #2670 : as said above by different people, it allows creating different atmospheres
3. #13639: this breaks the immersion so much. I'm sneaking, I'm not supposed to make any sound (or at least not as much as when walking/running). Very annoying in multiplayer

**Comment by DragonWrangler1** [comment]
**What I do**:  
I’m a very active modder, main developer of VoxelForge and make occasional shaders.

**Features**:
1. #2670. Why ? In order to have different colored nodes emit different colored light
2. Something along the lines of #13764. Why ? Because current liquids are too messy and annoying ,something like that would allow for games to better control how much flooding their liquids do.
3. Add a feature that allows games to set their own shaders (per game). Why ? this would allow for more diversity between games. Though still limited by what the engine allows.

**Comment by swagtoy** [comment]
1. Multiple worlds in the same world (no :airquote: """dimensions""" :airquote: at the top/bottom of the world hacks)
2. Better biome mapgen (i dont know anything about how v7 and stuff works but my buddy was complaining)
3.

**Comment by DragonWrangler1** [comment]
@swagtoy Just a little note. An issue does exist for at least the dimensions one.

**Comment by FreeLikeGNU** [comment]
What I do: I make mods that can work with several games (and make use of their existing content)
What I need: 
[Allow content to specify supported_games](https://github.com/minetest/minetest/issues/13851#issuecomment-1900730296) particularly on the subject of alternative depends.  Why? Optional depends does not really make clear what is needed to run a mod for a number of games.  For instance my mod can use `mobs` or `mcl_mobs` but it MUST have one OR the other.  I think that instead of declaring | in "`depends`", I would like an "`alt_depends`" (to not break existing mod.conf settings) in which to define mods that _must_ be present, but mutually exclusive. 
`alt_depends = mobs | mcl_mobs, flowers | mcl_flowers | hades_flowers`

**Comment by swagtoy** [comment]
@DragonWrangler1 Can you link it? I'll follow it if I can and read into it (Not sure why your message was hidden, considering you were "correcting" me.)

**Comment by DragonWrangler1** [comment]
@swagtoy #4428 I believe this is related ?

**Comment by Zughy** [comment]
I'd like to close this issue at the end of September, just to let people know

**Comment by FreeLikeGNU** [comment]
> I'd like to close this issue at the end of September, just to let people know

Seriously, marking contributions off topic because not conforming exactly to form?  How is that helping? People are taking time to respond and then being invalidated because their expression is slightly off.  It's not code being mainlined, it's commentary and discussion.  Please reconsider your moderation goals. Thank you.

**Comment by appgurueu** [comment]
> Seriously, marking contributions off topic because not conforming exactly to form? How is that helping? People are taking time to respond and then being invalidated because their expression is slightly off.

Please don't take this the wrong way. Essentially, it's just a way to collapse posts. The posts are not being invalidated.

This discussion is exactly about what it says in the title: Naming the most wanted features to help us prioritize. I'm following this topic for example and keeping my own (currently private) summary of what I plan to work on.

Discussion for these can then happen in the respective issues or PRs. Everything else, including this post, is essentially offtopic; the rules are clear about this. If we discuss the feature requests here, this thread will completely blow up.

P.S. I'm a bit surprised by some feature requests, including yours. Are these (is this) really the top feature(s) you need? Taking your request specifically as an example, you can essentially implement these requirements by declaring all these mods as optional dependencies, and shutting down the server / erroring with an informative message without any engine features if the requirements aren't met. This isn't ideal UX, but it works and users should be able to work with it. Would having some sort of engine feature really be that much of an improvement?

Given that you're working on mobs, I could personally imagine a few mobs-related features which I would probably consider more important...

**Comment by nauta-turbidus** [comment]
**What I do**:  
Maintainer of VoxeLibre.

**Features**:
1. https://github.com/minetest/minetest/issues/15176 This is needed to push Ambient Light over the line, and Ambient Light would be very versatile if it worked properly. I think we all understand the importance of it by now.
2. https://github.com/minetest/minetest/issues/5228 This could allow eg. biome generation based on a provided map. Think implementing a specific fantasy world with a specific map. If you need eg. mountain ranges in specific places (eg. for some Lord of the Test revamp), you'd have to do a bunch of Lua mapgen anyway, but in other cases custom biome system (or custom heat/humidity system, eg. based on a provided image and interpolation between values of the pixels) would be all it takes to make a map of a specific world, with different terrain each time. Currently the only workaround is to abuse min/max x/z in biome definition, and setting all biomes manually, which in some cases may be incredibly tedious.
3. https://github.com/minetest/minetest/issues/4758 All sorts of magical items for all sorts of games and mods could get a shiny animation. Fire (eg. torches?) carried in the inventory could get a shiny animation. Live animals carried in the inventory could be animated. All sorts of creative possibilities.

Aside of that, I'd support HUD meshes a lot, they'd be very useful for various projects, but it doesn't have a dedicated issue to my knowledge, and I decided not to have two of my own issues here.

**Comment by cx384** [comment]
**What I do**:
I make casual Minetest content, the [first mod](https://forum.minetest.net/viewtopic.php?t=9574) back in 2014, but I haven't been really active in recent years. However, I took this as an opportunity to think about things I'm missing in Minetest.

**Features**:
1. SSCSM #5393
  It was already planned in 2015 by #3440, but there has not been a lot of progress recently, and it's holding Minetest development back, since it is required/would be beneficial for a lot of features I would like to have.
  e.g. dynamic formspec changes, entity movement prediction, complex custom player physics, and other rendering/graphic effects, such that the server does not have to send them every time.

2. VAEs https://dev.minetest.net/TODO#Voxel_Area_Entities
  also see #4428
  I mainly want it to be able to create a game like [StarMade](https://en.wikipedia.org/wiki/StarMade) with Minetest.
  i.e. build/dig/move/walk-able spaceships, planets, asteroids, ...
  It could also be used to add smooth and diagonal movement to the digtron mod.

3. #15203
  In my opinion, connected textures look very nice. Most 2D games have nice texture tiling, but in three dimensions a fixed number of nodes for all possibilities is not feasible. Minetest is missing this for a long time, maybe I will make it happen eventually.

**Comment by farooqkz** [comment]
**What I do**
I contribute to Minetest CTF([link](https://github.com/MT-CTF/capturetheflag/graphs/contributors)). I am also doing some stuff around voice ML, Mozilla Common Voice(we get next why this matters!)

**Features**

 - SSCSM https://github.com/minetest/minetest/issues/5393. Imagine creating a game with which people contribute to Mozilla Common Voice. Contributions to MCV are either by saying or hearing. The hearing part is easy, you dynamically load the binary blob and then play it with the sound play API. But recording voice is not currently possible in Minetest. But if the server could tell the client to load a CSM which records audio and sends it to some server API, the record part of contributing to Mozilla Common Voice could be solved too. It could also have other applications. Such as in CTF I will suddenly shout "shangul! there is a red behind you!" while holding some key to record.
 - https://github.com/minetest/minetest/issues/14613 Not sure if I should include this here. But in PvP servers such as CTF, every single decisecond counts. Small amounts of lag still can be annoying. This could help reduce load on the server and therefore less lag or cheaper server costs.
 - https://github.com/minetest/minetest/issues/4637 This is really cumbersome others can chat in their own language but we can't chat in our native language. When you type Persian text into chat, nothing is shown. The same goes for Arabic as they are basically the same script. As for Hebrew, it is easier as to my knowledge Hebrew letters have the same connection no matter where in the word they are. I think some letters have different shapes when in the end of some word. But it's much easier than Arabic and Persian for sure. Also remember MCV? Two other ways to contribute to MCV is through writing and reviewing written sentences.

**Comment by Zughy** [comment]
Thanks everyone for participating. In general, I've tried to label at least one issue as "High priority" per comment; and, even if your issues didn't receive a "High priority" label, rest assured core devs definitely noticed (several issues received new comments). Closing!
